src/ast/mod.rs

```diff
@@ -942,12 +1096,47 @@ where
     }
 }
 
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub struct CorrespondingVar<Head, Leaf>
```

Comment:
Having a struct for CorrespondingVar is a great idea! 

Optional: the name is a little bit long, can we think of a simpler name? E.g., `WithWitness`?

Reply:
can't think of a better one